{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1a6b807b-f406-4736-81c6-124c00df7ca6",
   "metadata": {},
   "source": [
    "# Introduction to `scikit-allel`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8a8b07b-d485-4ba4-ac72-46faa82287af",
   "metadata": {},
   "source": [
    "The following Jupyter Notebook provides `Python` code to perform simple analyses of genome-wide single nucleotide polymorphism (SNP) data using the powerful `scikit-allel` package. As part of the your final project in Conservation Genetics you will use functions from this package to analyze empirical data; today, we will work with toy datasets to learn about the structure of data objects and how functions interact with them. To begin, you'll again need to copy this notebook and save an editable version to your own student directory. I also recommend spend some time reviewing `scikit-allel's` [documentation](https://scikit-allel.readthedocs.io/en/stable/), paying particular attention to what functions look like and how they operate."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e89d4a21-c970-4581-90eb-65e4bbdbf46c",
   "metadata": {},
   "source": [
    "**QUESTION 1**: Pick a function from the [\"Statistics and plotting\"](https://scikit-allel.readthedocs.io/en/stable/stats.html) section of `scikit-allel's` docs. What are its parameters, and what do they mean? Add a `Markdown` or `Raw` chunk below to write your answer."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ea7c165-1d29-4147-b30b-ff74b911e239",
   "metadata": {},
   "source": [
    "We will first load `scitkit-allel` and a few other packages:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "3f5d15a8-5c07-4165-8d78-5da8a5bab13c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import allel\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "np.set_printoptions(legacy='1.25')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b447936-86f8-4dad-aa12-77bb3d311599",
   "metadata": {},
   "source": [
    "## Basic data structures and functions\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e04e465-4af2-459e-83f5-1a4a1610236a",
   "metadata": {},
   "source": [
    "Let's get a feel for scikit-allel's `GenotypeArray` objects using a simple example. Here, we set up a `GenotypeArray` with two individuals (as columns) and three loci (as rows). Each integer in this array refers to an allele, where 0 indicates the reference allele, 1 the first alternate allele, 2 the second, etc. Any negative integer indicates missing data. Notice that we are again using `.` after the package name (`allel`) to specify where a function comes from. This time, however, we are creating something called a `class`, which assigns a particular set of variables and associated functions (technically \"methods\") to the object at hand:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "735471c2-4e49-4fdf-8fe3-d423d3349ee0",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = allel.GenotypeArray([[[0, 0], [0, 0]],\n",
    "                         [[0, 0], [0, 1]],\n",
    "                         [[0, 0], [1, 1]],\n",
    "                         [[0, 1], [1, 1]],\n",
    "                         [[1, 1], [1, 1]],\n",
    "                         [[0, 0], [1, 1]],\n",
    "                         [[0, 1], [1, 0]],\n",
    "                         [[0, 1], [-1, -1]],\n",
    "                         [[-1, -1], [-1, -1]]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "165f20f2-e57c-4d47-a7df-78e69e4b5524",
   "metadata": {},
   "source": [
    "This is what that object looks like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b61aec02-18d0-4a0f-9d3e-32826068e129",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAs2D\"><span>&lt;GenotypeArray shape=(9, 2, 2) dtype=int64&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">0</th><th style=\"text-align: center\">1</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/1</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">1/1</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"3\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">6</th><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">1/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">7</th><td style=\"text-align: center\">0/1</td><td style=\"text-align: center\">./.</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">8</th><td style=\"text-align: center\">./.</td><td style=\"text-align: center\">./.</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<GenotypeArray shape=(9, 2, 2) dtype=int64>\n",
       "0/0 0/0\n",
       "0/0 0/1\n",
       "0/0 1/1\n",
       "...\n",
       "0/1 1/0\n",
       "0/1 ./.\n",
       "./. ./."
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6cc4833-2852-4dc8-86a4-31fe47292d79",
   "metadata": {},
   "source": [
    "Our GenotypeArray object \"g\" has attributes reflecting its dimensions, its number of variants, ploidy and sample size. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "c875a018-7442-4caa-96e8-9e11cd2e2638",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g.ndim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e622a54e-1fac-4515-bd44-9afcd4388617",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9, 2, 2)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "92f96a55-88d0-4c8e-89e5-449d97bd2810",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g.n_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "17c27d87-7588-4d1d-a023-da3882f852bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g.ploidy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d47d3a8f-20bc-4584-b2da-387e302361dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g.n_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9616e6bc-c2fa-4b84-9658-d21d023a07eb",
   "metadata": {},
   "source": [
    "With this object, we can begin to get a feel for `scikit-allel's` functions for describing diversity and divergence. For example, we can calculate observed heterozygosity, generating an array for each locus:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "4a5f0041-d4a2-4c47-a50b-8f67324124c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0. , 0.5, 0. , 0.5, 0. , 0. , 1. , 1. , nan])"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "het_obs = allel.heterozygosity_observed(g)\n",
    "het_obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f13d6ce-d7b5-413c-a0fc-88fc0aac1291",
   "metadata": {},
   "source": [
    "Using the `namean()` function in `numpy` to ignore `NA` values, we can calculate the genome-wide average:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "c7bb95da-47b3-4263-b593-0197873149d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(0.375)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.nanmean(het_obs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8e01f84-066c-4b6c-ace1-7a89905bc5df",
   "metadata": {},
   "source": [
    "A natural next step is to see whether these freququencies are a deviation from Hardy-Weinberg Equilibrium. To do so, we first calculate allele frequencies for each locus, using the `count_alleles().to_frequencies()` function. (Note that this is an example of of an object having a particular method—in this case, a method unique to the `GenotypeArray()` class.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "cd869928-bcec-43e7-be24-ca55c64892f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1.  , 0.  ],\n",
       "       [0.75, 0.25],\n",
       "       [0.5 , 0.5 ],\n",
       "       [0.25, 0.75],\n",
       "       [0.  , 1.  ],\n",
       "       [0.5 , 0.5 ],\n",
       "       [0.5 , 0.5 ],\n",
       "       [0.5 , 0.5 ],\n",
       "       [ nan,  nan]])"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af = g.count_alleles().to_frequencies()\n",
    "af"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c6a6b77-7cbf-4e99-87da-3bcaf4d2130e",
   "metadata": {},
   "source": [
    "In this array, each row is a locus, and the columns 0 and 1 and refer to reference and alternate alleles (though note that there can be more columns if the locus is not diallelic). Next, we can use these data to calculate expected heterozygosity:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "8b16316c-8bdb-4535-94fa-f1a34e3d86ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.   , 0.375, 0.5  , 0.375, 0.   , 0.5  , 0.5  , 0.5  ,   nan])"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allel.heterozygosity_expected(af, ploidy=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfa7ba37-b2a0-47c9-909a-d95ef7c31c61",
   "metadata": {},
   "source": [
    "Looks different to me!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0084e6da-fe49-4f48-b55b-8c4cb9e327e3",
   "metadata": {},
   "source": [
    "**QUESTION 2**: What is the length of object `af`? Add a chunk of code below to answer this question."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd1a1acd-e364-497b-bd3d-75b743886755",
   "metadata": {},
   "source": [
    "We can select a single site using brackets—for example, the following line of code gives us the frequencies of the two alleles at the first site in the genome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "c012edbc-af0d-4d64-a2d3-c9bb6d9b1a39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.75, 0.25])"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c8debe7-a28b-43e4-82cd-a4bdbb8f1bd9",
   "metadata": {},
   "source": [
    "We can use the following function to calculate observed heterozygosity at one or more SNPs, again using brackets to select a particular location in the genome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "d684f6d0-37e8-4a3b-8c9a-8b64d9dce35f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(0.0)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_o = allel.heterozygosity_observed(g)[3] #this selects the third variant\n",
    "h_o"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2dc5670-1dd7-4b6d-8b4e-9ad204678a19",
   "metadata": {},
   "source": [
    "The function `allel.heterozygosity_expected` works similarly, albeit for the frequency data, not the `GenotypeArray`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "56046114-8bc4-4420-85a3-84273da18005",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(0.5)"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_e = allel.heterozygosity_expected(af, ploidy=2)[3] #this selects the third variant\n",
    "h_e"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7da79a7f-eb65-412c-9934-dfefdc882cc4",
   "metadata": {},
   "source": [
    "**QUESTION 3**: Recall that the inbreeding coefficient can be calculated as $F = 1 - \\frac{H_o}{H_e}$. Add a chunk below to determine the inbreeding coefficient for the *second* variant. What can you infer from the answer? (This will take more than one line of `Python`.)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
