{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b45392fe-a6c5-4ba2-8c02-666ee3fa3942",
   "metadata": {},
   "source": [
    "# BIOB480/BIOE548 Final Project"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "116f22b0-a88d-4304-affc-a51879de061b",
   "metadata": {},
   "source": [
    "To conclude the semester, students will work independently to analyze empirical genetic or genomic data from a population of wild plants or animals. Specifically, each of you will select a published dataset and population genetic metric, read the abstract of the associated scientific article, load the data into a Jupyter Notebook, and perform simple analyses using the Python package scikit-allel. Learning outcomes include gaining familiarity with Python and scriptable anlayses of genomic data more broadly, learning how sequence data and genotypes are encoded into data objects. The project is worth 25 points and must be submitted as an `.ipynb` notebook in your local directory on Tempest. Of these points, 10 are for completeness, while 15 are earned by answers to the questions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fefe9d78-07da-4ce9-844f-6a42b69cca92",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6a0693c-c458-4158-8c4d-8cf0386011f0",
   "metadata": {},
   "source": [
    "As before, we're going to load several `Python` libraries to facilitate analysis of genotype data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d4eaef08-582d-48cb-8c18-37d60da32438",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import allel\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "np.set_printoptions(legacy='1.25')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08ebada3-7629-4ebe-abae-fb2b0540dedb",
   "metadata": {},
   "source": [
    "## Selecting a Dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52c5fdd2-3eb2-46e6-8f82-038cdf34de2b",
   "metadata": {},
   "source": [
    "To start, visit the [linked GitHub repository](https://github.com/elinck/conservation_genetics) and read the README.md file on the main page. There, I introduce a set of empirical conservation genetics datasets, with links to compressed Variant Call Format (`.vcf`) files and their associated papers. I've already moved these data to the subdirectory `/biob-480-genetics/course-materials/data/` on Tempest, so all you need to do is decide on study taxon of interest. Once you've done so, read at least the abstract of the linked paper, and review all figures and tables. This will inform your answer to **QUESTION 1** below."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1896044-5631-428b-abdd-15a02a429e48",
   "metadata": {},
   "source": [
    "**QUESTION 1**: Create a new code chunk immediately below this field. Click inside the box and select `Markdown` from the dropdown field that currently reads `Code` above. Markdown is a simple formatting language for text——surrounding words with a pair of asterices (e.g., `*text*`) makes them *italic*, while two pairs of asterices (`**text**`) makes them **bold**. Clicking the \"run\" button will then format all text in the code chunk according to `Markdown`'s syntax (a handy cheatsheet of other rules [is available here](https://www.markdownguide.org/cheat-sheet/). Though the formatting is not important, you'll use `Markdown` chunks whenever you need to respond to a question with written text. Here, I'd like you to state which dataset you selected, and briefly summarize the question, genetic market type, and main findings in your own words. Why did it interest you?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28abd460-75b0-4180-80d9-f5fd9ef46c9e",
   "metadata": {},
   "source": [
    "## Loading and examining variant data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c694d76f-2b3a-46c6-b57d-86cdde43a3b0",
   "metadata": {},
   "source": [
    "Next up, we're going to pivot to looking at empirical data: specifically, the SNP data associated with your project. These data are in Variant Call Format, and can be found as `*.vcf.gz` in the tutorial folder. In the example below, we'll load a test dataset present in the `course-materials/` subdirectory. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "93e99357-9b23-4c1b-a095-a539450c62eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = \"/home/group/biob-480-genetics/course-materials/data/iris.vcf.gz\"\n",
    "s = allel.read_vcf(file, fields='*')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8acb88de-1961-45ca-8066-e95200ce964f",
   "metadata": {},
   "source": [
    "The numpy array object `s` has a set of keys representing aspects of the `.vcf` file. We can list these with the `.keys()` function, sorting by name:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a8ebad97-9d5d-46dd-bbe2-f4888bc8f6d1",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 's' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[1]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[38;5;28msorted\u001b[39m(\u001b[43ms\u001b[49m.keys())\n",
      "\u001b[31mNameError\u001b[39m: name 's' is not defined"
     ]
    }
   ],
   "source": [
    "sorted(s.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bfee6c6-96f0-487d-b62e-349ca8a151bd",
   "metadata": {},
   "source": [
    "Here, anything begining with \"callset\" is from the INFO field of the .vcf, while \"variants\" refers to data associated with each SNP (or site, if you have invariant alleles in your file). (Some of this information, e.g. depth, is redundant, depending on your SNP caller.) To examine any of these in greater detail, we can select the relevant index. For example, to look at the position of these SNPs (relative to the reference genome or sequence they were aligned to), we can select `variants/POS` key. Here, we use brackets with a `String` inside to select a column with a particular label:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb528acf-d833-4f93-a38e-aca75d0a2c75",
   "metadata": {},
   "outputs": [],
   "source": [
    "s['variants/POS']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db6685db-d595-4550-b0e1-7c149f14b5e1",
   "metadata": {},
   "source": [
    "Similarly, we can list all sample IDs, which in most cases will be filenames from one of the various bioinformatics processing steps used to generate the final `.vcf` file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ccca5c1-0e72-4df0-bd16-945747a9f7a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "s['samples']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95c7e900-3df3-48b5-9726-17ea540fa330",
   "metadata": {},
   "source": [
    "**QUESTION 2**: How many samples are in your dataset? To answer this, you may add a new code chunk below and comment directly within it next to your script with a hashtag (`#`), or add a `Markdown` chunk above your code chunk to add written text (e.g., \"There are 102 samples in my dataset:\", followed by a separate line with the code you used to determine your answer)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79171a6b-2991-407e-9c68-8c50409efd09",
   "metadata": {},
   "source": [
    "Besides simply looking at these data, these keys can be used to filter SNPs based on various criteria. First, though, you may want to know how many SNPs you acutally have. We can evaluate this by taking the first entry from the shape attribute of the array (i.e., its length): "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "706041d1-b65f-48f1-ab5f-c0a8a098e637",
   "metadata": {},
   "outputs": [],
   "source": [
    "s['calldata/GT'].shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31ea15c6-7b75-4703-a634-8bffd5fca2d7",
   "metadata": {},
   "source": [
    "Next, let's take the actual genotype calls from the vcf-turned-numpy-array and turn it into the GenotypeArray object we're familiar with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56f7e86a-7417-4841-879c-04a2b15cfd94",
   "metadata": {},
   "outputs": [],
   "source": [
    "gt = s['calldata/GT']\n",
    "gt = allel.GenotypeArray(gt)\n",
    "gt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5faeed1-9520-451f-affc-0e0f22888bb3",
   "metadata": {},
   "source": [
    "**QUESTION 3**: How many loci are in your dataset?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f5076b8-d774-4a76-b4cb-8183bf13969f",
   "metadata": {},
   "source": [
    "## Metrics of Genetic Diversity"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f468b758-b458-47bc-9e6d-23819dac92a2",
   "metadata": {},
   "source": [
    "Let's calculate the frequencies of different alleles (SNPs) using the `count_alleles().to_frequencies()` function introduced above, saving them to a new object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0d9c2122-c2b8-45e1-82f0-3d78c94eef54",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'gt' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[3]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m af_gt = \u001b[43mgt\u001b[49m.count_alleles().to_frequencies()\n\u001b[32m      2\u001b[39m af_gt\n",
      "\u001b[31mNameError\u001b[39m: name 'gt' is not defined"
     ]
    }
   ],
   "source": [
    "af_gt = gt.count_alleles().to_frequencies()\n",
    "af_gt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cf9ccd3-dfed-4b37-b069-f3a789e48091",
   "metadata": {},
   "source": [
    "We can select a single site using brackets—for example, the following line of code gives us the frequencies of the two alleles at the second site in the genome (recall that `Python` uses 0-indexing, i.e. the first site is `[0]`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d920c7d-7d17-44b3-9ed3-267a1ee8d6ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "af_gt[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bde66e3-a772-45df-b91c-40e6a7c8a597",
   "metadata": {},
   "source": [
    "We can use the following function to calculate observed heterozygosity at each SNP in the dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23385aa9-77b0-4580-8570-62b9b92cfe59",
   "metadata": {},
   "outputs": [],
   "source": [
    "h_o = allel.heterozygosity_observed(gt)\n",
    "h_o"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bd59c19-ce1e-4145-b112-5add41ce96c2",
   "metadata": {},
   "source": [
    "**QUESTION 4**: What is the observed heterozygosity at the 32nd site in your data?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "003cdc10-9b03-4e40-822a-7d7b30c86fed",
   "metadata": {},
   "source": [
    "The function allel.heterozygosity_expected works similarly:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46a377dd-f90d-488e-b47f-a96d3e98abe2",
   "metadata": {},
   "outputs": [],
   "source": [
    "h_e = allel.heterozygosity_expected(af_gt, ploidy=2)\n",
    "h_e"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62edb5ca-400d-4316-86bc-13754514c269",
   "metadata": {},
   "source": [
    "Recall that $F = 1 - \\frac{H_o}{H_e}$. We therefore have enough to calculate the value of the inbreeding coefficient for any locus:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7751782f-a97c-4bc8-8b0e-84fd0da04f4b",
   "metadata": {},
   "outputs": [],
   "source": [
    " allel.inbreeding_coefficient(gt)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c26ccc5-7730-4e02-b648-40a236c809c8",
   "metadata": {},
   "source": [
    "**QUESTION 5**: Calculate $H_e$, $H_o$, and $F$ for SNPs at positions 6, 19, and 72. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8ec6469-5d22-40ea-85aa-f03346b4c898",
   "metadata": {},
   "source": [
    "Another basic descriptive task in conservation and population genetics is assessing the diversity across loci in a set of samples. Our empiricial data have already been filtered down to variant sites, so we'll work with the simple genotype array `g` from last lesson to get a hang of how a few functions work. To begin, we need to load it again:  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df6868c6-31b8-4124-9572-744c81411165",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = allel.GenotypeArray([[[0, 0], [0, 0]],\n",
    "                         [[0, 0], [0, 1]],\n",
    "                         [[0, 0], [1, 1]],\n",
    "                         [[0, 1], [1, 1]],\n",
    "                         [[1, 1], [1, 1]],\n",
    "                         [[0, 0], [1, 1]],\n",
    "                         [[0, 1], [1, 0]],\n",
    "                         [[0, 1], [-1, -1]],\n",
    "                         [[-1, -1], [-1, -1]]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04735c5d-830d-4a31-9026-75557ee2fb73",
   "metadata": {},
   "source": [
    "We next need to convert our genotype array into counts of each allele at each position. (Note that the maximum value will be 2N=4.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efb0a756-5aa8-4e51-913b-c93481bd8681",
   "metadata": {},
   "outputs": [],
   "source": [
    "g_ac = g.count_alleles()\n",
    "g_ac"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa3547ed-92e6-4470-91fe-83f4336fbc93",
   "metadata": {},
   "source": [
    "There are 9 sites in this allele count array. Let's imagine they correspond to positions 2,4,7,14,15,18,19, and 27 in a reference genome that is 32 nucleotides long. We can indicate this with an array assigned to `pos`, and then calculate $\\pi$ using the function `allel.sequence_diversity()`. (Note that positions 1 through 31 indicate the stretch of the genome these SNPs are taken from, including invariant sites.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "df4d06fd-c38a-49b6-bfad-900dddc8e751",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'allel' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[5]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m pos = [\u001b[32m2\u001b[39m, \u001b[32m4\u001b[39m, \u001b[32m7\u001b[39m, \u001b[32m14\u001b[39m, \u001b[32m15\u001b[39m, \u001b[32m18\u001b[39m, \u001b[32m19\u001b[39m, \u001b[32m25\u001b[39m, \u001b[32m27\u001b[39m]\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m pi = \u001b[43mallel\u001b[49m.sequence_diversity(pos, g_ac, start=\u001b[32m1\u001b[39m, stop=\u001b[32m31\u001b[39m)\n\u001b[32m      3\u001b[39m pi\n",
      "\u001b[31mNameError\u001b[39m: name 'allel' is not defined"
     ]
    }
   ],
   "source": [
    "pos = [2, 4, 7, 14, 15, 18, 19, 25, 27]\n",
    "pi = allel.sequence_diversity(pos, g_ac, start=1, stop=31)\n",
    "pi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8829a6ca-8ec0-4dc5-9dc5-8996f3477afa",
   "metadata": {},
   "source": [
    "Another useful metric of diversity is the mean number of pairwise differences for a particular locus. For SNP data, this is equivalent to comparing each allele in each individual and scoring them as 0 (no difference) or 1 (different alleles). We can calculate this for all SNPs in our empirical data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8aed82cd-f1eb-4c05-a2e5-87aa4e72c53b",
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_ac = gt.count_alleles()\n",
    "dif = allel.mean_pairwise_difference(gt_ac)\n",
    "dif"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee4046e8-57e6-417e-9384-5f178c051419",
   "metadata": {},
   "source": [
    "**QUESTION 6**: Remember that basic Python 3 function `max()` calculates the maximum value of a given array. What is the largest mean pairwise difference value in your dataset?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56c920c1-d73d-47df-8654-4144602173f1",
   "metadata": {},
   "source": [
    "## Linkage Disequilibrium "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebc4c51e-7958-4840-9631-63298b31033b",
   "metadata": {},
   "source": [
    "Recall that linkage disequilibrium—the association of alleles at different loci at great—is typically measured by the metric $D$. A related metric, $r$, can be interpreted as the [correlation between two diploid genotypes](doi.org/10.1534/genetics.108.093153). We can calculate this for our empirical data after first manipulating our genotype array to instead display the number of non-reference alleles (i.e., we are coding diploid genotypes for each individual at each SNP by the number of alleles they have that are different than the reference genome):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4908552-a102-4379-a5a0-14d2fa427fb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_subset = gt[1:50]\n",
    "gt_an = gt_subset.to_n_alt(fill=-1)\n",
    "gt_an[41]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85cac4a6-9795-456f-93fc-1b54ab4181a4",
   "metadata": {},
   "source": [
    "A seperate function returns a matrix of pairwise (sample to sample) LD values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49c1038b-549b-4da4-9c06-7e8edb44c62e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ld = allel.rogers_huff_r(gt_an)\n",
    "ld"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7d5cb92-5ec3-4a41-bf5b-b4dd6f843b75",
   "metadata": {},
   "source": [
    "We can plot these using a built in function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e336fad2-cc39-4c25-bb89-0d5bbb2cb3ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "allel.plot_pairwise_ld(ld, colorbar=True, ax=None, imshow_kwargs=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03df57e9-0232-4c37-871c-f3b351959b9c",
   "metadata": {},
   "source": [
    "**QUESTION 7**: What is the largest value of $r$ in your dataset? Is it low or high?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b701acc5-a554-406b-b762-9c0cd9359c78",
   "metadata": {},
   "source": [
    "Let's plot a histogram of these same data. What do you notice?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "84a96295-b146-4067-adc0-8be3cc4908f9",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'plt' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[8]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mplt\u001b[49m.hist(ld, bins=\u001b[33m'\u001b[39m\u001b[33mauto\u001b[39m\u001b[33m'\u001b[39m) \u001b[38;5;66;03m# arguments are passed to np.histogram\u001b[39;00m\n\u001b[32m      2\u001b[39m plt.title(\u001b[33m\"\u001b[39m\u001b[33mHistogram of pairwise $r$ values\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m      3\u001b[39m plt.show()\n",
      "\u001b[31mNameError\u001b[39m: name 'plt' is not defined"
     ]
    }
   ],
   "source": [
    "plt.hist(ld, bins='auto') # arguments are passed to np.histogram\n",
    "plt.title(\"Histogram of pairwise $r$ values\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40a2f9be-0114-4298-83b7-a4b56d470715",
   "metadata": {},
   "source": [
    "## Population Subdivision"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc077ccc-2c6c-41a8-ae63-5d0242505c4b",
   "metadata": {},
   "source": [
    "Most (all?) of your datasets include samples from multiple, geographically isolated populations. Properly analyzing these data would require linking sample IDs to specific populations, which should be possible, but is difficult to standardize across datasets for our purposes. Therefore, you can decide to choose one of two options to estimate $F_{ST}$ or your dataset: 1) an arbitrary assignment of samples to \"subpopulations\"; or 2) accurately assigning individuals to subpopulations on the basis of the source paper and / or other files in the [Dryad repository linked on GitHub](https://github.com/elinck/conservation_genetics)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dea479de-ce15-4400-8f7f-b7d1620be6b9",
   "metadata": {},
   "source": [
    "For the first option, we will arbitrarily divide our samples and associated alelle count arrays into two groups:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77e7b57a-e922-4624-a31b-5917aa36e995",
   "metadata": {},
   "outputs": [],
   "source": [
    "first_half = int(gt.n_samples/2) # select a sample number half-way through the total number of samples\n",
    "second_half = int(gt.n_samples) # select the final sample\n",
    "subpops = list(range(first_half)), list(range(first_half+1,second_half)) # assign subpopulations as a list of sample numbers\n",
    "gt_ac1 = gt.count_alleles(subpop=subpops[0]) # calculate allele counts for the first \"subpopulation\"\n",
    "gt_ac2 = gt.count_alleles(subpop=subpops[1]) # calculate allele counts for the second \"subpopulation\"\n",
    "num, den = allel.hudson_fst(gt_ac1, gt_ac2) # calculate the numerator and denominator for FST for each SNP\n",
    "fst = np.sum(num) / np.sum(den) # calculate average FST by summing over all SNPs\n",
    "fst"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a1ad834-352f-499a-8036-773761caf00a",
   "metadata": {},
   "source": [
    "**QUESTION 8**: What is the genome-wide average $F_{ST}$ value for your data calculated in this way? Why is it so low?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "028e26cc-96d6-43ca-88ad-cda5ccfc92e3",
   "metadata": {},
   "source": [
    "Recall that we loaded our `.vcf` files as an object called `s`. Among its attributes is a list of samples—let's display the first 5 sample names as an array:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c89ca7f-0197-4ec8-9f1b-02d4ac9c37db",
   "metadata": {},
   "outputs": [],
   "source": [
    "s['samples'][0:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f58daec-1950-438d-9b7e-6ae3ee31a6ac",
   "metadata": {},
   "source": [
    "By inspecting this file and comparing it to background information, we can theoretically assign individuals to actual subpopulations by their index. If possible, modify the following code to do so, calculating $F_{ST}$ between two of separate populations in your study. (If there are no lists of sample IDs / localities available, you may select another random subset.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c35bae10-5c49-445b-a841-5d3d879e32dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "subpops2 = [[0,1,2,3,4],[90,91,92,93,94]] # this would assign individuals 0-4 and 90-94 to two different subpopulations; edit as needed\n",
    "subpops2\n",
    "gt_ac1_v2 = gt.count_alleles(subpop=subpops2[0]) \n",
    "gt_ac2_v2 = gt.count_alleles(subpop=subpops2[1]) \n",
    "num, den = allel.hudson_fst(gt_ac1_v2, gt_ac2_v2) \n",
    "fst = np.nansum(num) / np.nansum(den) \n",
    "fst"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30e288cf-bd15-473d-83c4-91478bef1b13",
   "metadata": {},
   "source": [
    "**QUESTION 9**: Is this $F_{ST}$ value lower or higher than your answer to **QUESTION 8**?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2142d249-ec94-44e6-9355-4961b76bbf52",
   "metadata": {},
   "source": [
    "Let's also calculate *per variant* $F_{ST}$---this way, we can see whether certain SNPs / loci are outliers (and thus potentially under selection). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "847db34c-2c14-48c1-b6ff-060ceacf428c",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'np' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[10]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mnp\u001b[49m.seterr(divide=\u001b[33m'\u001b[39m\u001b[33mignore\u001b[39m\u001b[33m'\u001b[39m, invalid=\u001b[33m'\u001b[39m\u001b[33mignore\u001b[39m\u001b[33m'\u001b[39m) \u001b[38;5;66;03m# allow division by zero\u001b[39;00m\n\u001b[32m      2\u001b[39m fst = num / den \n\u001b[32m      3\u001b[39m fst = fst[~np.isnan(fst)] \u001b[38;5;66;03m# drop nan values\u001b[39;00m\n",
      "\u001b[31mNameError\u001b[39m: name 'np' is not defined"
     ]
    }
   ],
   "source": [
    "np.seterr(divide='ignore', invalid='ignore') # allow division by zero\n",
    "fst = num / den \n",
    "fst = fst[~np.isnan(fst)] # drop nan values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d56d0cfb-a8e1-4c02-8ee8-48d6d53e2eeb",
   "metadata": {},
   "source": [
    "As with our LD values, we can plot a histogram of per-SNP $F_{ST}$: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d081317-0888-44ad-9cfe-cb236b935f8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(fst, bins='auto')\n",
    "plt.title(\"Histogram of per-SNP $F_{ST}$ values\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "642c3f61-0e32-4940-919e-2ebf35ff3c2d",
   "metadata": {},
   "source": [
    "**QUESTION 10**: What is the maximum $F_{ST}$ value in your dataset? Be precise (i.e., use a function and don't just eyeball the plot). For an extra challenge, determine its position in your dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cec74f7-8d86-4f5e-a33b-9087ca8c5897",
   "metadata": {},
   "source": [
    "## Principal Component Analysis "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29af77bf-359f-4eb3-a15c-d5325d15dd2b",
   "metadata": {},
   "source": [
    "To conclude, we'll perform Principal Component Analysis (PCA) on our data. Though we did not explicitly cover PCA in class, it is a ubiquitous approach to extracting patterns from highly multivariate datsets (like genotypes from thousands of SNPs and hundreds of samples) that involves placing samples on new, \"synthetic\" axes that explain the majority of variation in the data. A brief introduction to PCA [is here](https://en.wikipedia.org/wiki/Principal_component_analysis); arguably its most famous use in population genetics is [this 2008 paper in Nature](https://www.nature.com/articles/nature07331) by John Novembre and colleagues. PCA is most useful when we know the identity of each plotted sample, but even without going through the necessary steps to do so with our own data, the patterns it reveals can be informative."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba5137d9-0813-4b5d-9251-9d0f0b038e0a",
   "metadata": {},
   "source": [
    "`scikit-allel` has a built-in PCA function, which is straightforward to implement once we are sure our data are complete (i.e., have no missing values). To do so, we'll start by counting alleles as before:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8b5d4b5-5b20-4b67-a9bd-3175634f972d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ac_sub = gt.count_alleles() # count alleles\n",
    "ac_sub"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "057ac96a-c410-4d16-9801-87da66454be9",
   "metadata": {},
   "source": [
    "Because some of the datasets include invariant sites, which will cause problems with PCA, we need to drop them from the object. To do so, we will combine `Numpy`'s `np.nonzero` function with the `is_variant` method of `allel`'s genotype datastructures. (The \"0\" below makes sure we end up with a single array; don't worry about it.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84b78c71-f258-4e7f-a0e9-a2798ff324df",
   "metadata": {},
   "outputs": [],
   "source": [
    "indices = np.nonzero(ac_sub.is_variant())[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9070a7b-c138-44da-9b6a-4fa73cd32b74",
   "metadata": {},
   "source": [
    "Once we have these values, we can use the `take` method of genotype datastructures to subset them:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3c48988-9ea3-4ba1-ae34-924f1704eb1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "indices = np.nonzero(ac_sub.is_variant())[0] # identify only variant sites\n",
    "genotypes_pca = gt.take(indices, axis=0) # subset genotypes to variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a4e9437-6caa-4966-a9fa-d931b5f06879",
   "metadata": {},
   "outputs": [],
   "source": [
    "before = len(gt)\n",
    "after = len(genotypes_pca)\n",
    "SNPs_dropped = before - after\n",
    "SNPs_dropped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4e9f226-6084-4eee-a94d-0c439f9f6dc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "(Depending on your dataset, this value may be zero.)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5c70c3c-50a7-46e1-89d4-af2e1c45bd1f",
   "metadata": {},
   "source": [
    "Our last data preparation step is to transform our data to an array counts of alternate alleles—i.e., alleles that are different than the reference sequence reads were aligned to during sequence assembly:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0d8f830b-a2cd-4400-abf4-238d103ead51",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'genotypes_pca' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[12]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m gn = \u001b[43mgenotypes_pca\u001b[49m.to_n_alt()\n\u001b[32m      2\u001b[39m gn\n",
      "\u001b[31mNameError\u001b[39m: name 'genotypes_pca' is not defined"
     ]
    }
   ],
   "source": [
    "gn = genotypes_pca.to_n_alt()\n",
    "gn"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52a93891-f45b-4e25-81b7-b791b7c7d3af",
   "metadata": {},
   "source": [
    "Now it's time to run PCA itself. Note that we are saving information on the model used to run it in the second variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea43aeec-d9dc-4215-8669-f3fa41430dcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "coords, model = allel.pca(gn) # run pca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f86c928-a1b2-45ec-b821-4e4b6eb071b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(coords[:, 0], # position on the first principal component of the observations\n",
    "            coords[:, 1], alpha=0.7);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01ce15fa-74b0-4924-8717-2d19a9da012e",
   "metadata": {},
   "source": [
    "**QUESTION 11**: Interpret this PCA. Do you see distinct clusters, suggesting groups of samples that are closely related to each other, or do you see largely continuous variation? What biological processes relevant to your study taxon might help explain these patterns? Do your results match conclusions in the paper?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46cb960e-84b0-4ff2-b90a-ffc8163b7352",
   "metadata": {},
   "source": [
    "That's the end of your brief tour of `scikit-allel`! We have, of course, only scratched the surface of what you can do with sequence data in Python. If you are motivated, feel free to add new code chunks below and perform an addititional analysis or data visualization on your dataset using any resource you wish for help. If your work is successful and you document it, I'll give you extra credit (**QUESTION 12**)!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad9bb22a-3571-4272-b36f-77ff664f3e37",
   "metadata": {},
   "source": [
    "**QUESTION 13**: Please fill out an evaluation for this course using [this link (for undegraduates)](https://montana.campuslabs.com/eval-home/direct/8519695) or [this link (for graduate students)](https://montana.campuslabs.com/eval-home/direct/6550795). Confirm you have done so in a new `Markdown` chunk below.\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db1631fb-e44b-4237-babc-67bf145c5e52",
   "metadata": {},
   "source": [
    "**QUESTION 14**: Did you enjoy the programming activities in this class? What could be done to improve them for future semesters?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae3f12c5-9580-4de9-815e-d7e6c1516958",
   "metadata": {},
   "source": [
    "**QUESTION 15:** What was most surprising, exciting, or disappointing to you about this course?"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
